src/remote/fs/fs.go

```diff
@@ -0,0 +1,210 @@
+// Package fs provides an io/fs.FS implementation over the remote API
+package fs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	iofs "io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/bazelbuild/remote-apis-sdks/go/pkg/client"
+	"github.com/bazelbuild/remote-apis-sdks/go/pkg/digest"
+	pb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"google.golang.org/protobuf/proto"
+)
+
+type Client interface {
+	ReadBlob(ctx context.Context, d digest.Digest) ([]byte, *client.MovedBytesMetadata, error)
+}
+
+// fs is an io/fs.FS implemented on top of a REAPI directory. This will download files on demand, as they are needed.
+type fs struct {
+	c    Client
+	root *pb.Directory
+}
+
+// New creates a new filesystem on top of the given proto, using client to download files on demand.
+func New(c Client, root *pb.Directory) iofs.FS {
+	return &fs{
+		c:    c,
+		root: root,
+	}
+}
+
+// Open opens the file with the given name
+func (fs *fs) Open(name string) (iofs.File, error) {
+	return fs.open(".", filepath.Clean(name), fs.root)
+}
+
+func (fs *fs) open(path, name string, wd *pb.Directory) (iofs.File, error) {
+	name, rest, hasToBeDir := strings.Cut(name, string(filepath.Separator))
+	// Must be a dodgy symlink that goes past our root.
+	if name == ".." || name == "." {
+		return nil, os.ErrNotExist
+	}
+
+	for _, d := range wd.Directories {
+		if d.Name == name {
+			dg, err := digest.NewFromProto(d.Digest)
```

Comment:
FWIW you can probably get away with `NewFromProtoUnvalidated`, we shouldn't ever have invalid digest protos hanging around the place. Also fine as is tho